src/vdaf.rs

```diff
@@ -248,6 +250,24 @@ pub trait Aggregator<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>: Vda
         agg_param: &Self::AggregationParam,
         output_shares: M,
     ) -> Result<Self::AggregateShare, VdafError>;
+
+    #[cfg(feature = "experimental")]
+    /// Adds noise to an aggregate share to achieve differential privacy.
+    /// The `dist` distribution contains the relevant DP parameters.
+    /// The noise is drawn from `dist` and then scaled by `num_measurements`
+    /// if necessary, in order to enforce the desired privacy level.
+    ///
+    /// TODO(tholop): VDAFs that don't want DP can make it a no-op.
+    ///               How about VDAFs that sometimes want DP and sometimes not?
+    ///               Leave it up to them how they implement their `dist` parameter,
+    ///               potentially with a `None` variant?
+    fn add_noise_to_agg_share<D: Dist>(
```

Comment:
Ahh interesting point. One option would be to move `Dist` to the trait instead:
```rust
pub trait CentralDpAggregator<D: Distribution>: Aggregator {
    fn add_noise_to_agg_share( ... dist: &D) -> ....
}
```

that way you have to explicitly implement `CentralDpAggregator<DiscreteGaussian>` for your VDAF.

Reply:
Here are a few scenarios for how we could handle this:

* Distributions are passed as a trait object, as described above. We could use `Any::downcast()` as a more idiomatic way to check the distribution's type, instead of checking codepoints. We could also consider providing a no-op distribution that doesn't read the seed stream and always returns zeros. This could make aggregator code paths a little simpler, as they could unconditionally call `add_noise_to_agg_share()`.
* The distribution and its parameters are passed as an enum, and each VDAF returns an error if an inappropriate distribution is chosen (i.e. a mechanism for categorical data is used with Prio3Sum).
* `add_noise_to_agg_share()` is moved to some new trait, called `AggregatorWithNoise`,  `NoiseAdder`, or similar, with `Aggregator` as a super trait. The noise distribution is a type parameter of this trait. Then, VDAFs that support DP implement it with the trait's type parameter equal to whichever noise distributions are appropriate. This rules out the need for a runtime check.

Reply:
I like option 3 :)

Reply:
One other point in favor of a separate trait: with the current state of this PR, the shape of the Aggregator trait depends on the `experimental` feature. Thus, compilation would break if any foreign VDAF implementation, written assuming the feature is off, either turns it on for some other reason (Poplar1, Prio2, fixed point bounded L2 norm Prio3 instance etc.) or unintentionally turns it on through transitive dependencies that enable it. If we moved the new method to a separate trait, then the conditional compilation would only change the presence or absence of entire items, which is much easier to work with.

Reply:
I moved `add_noise_to_agg_share()` to a new `AggregatorWithNoise` trait.